crates/stages/src/stages/headers.rs

```diff
@@ -167,6 +166,29 @@ impl<D: HeaderDownloader, C: Consensus, H: HeadersClient, S: StatusUpdater>
         Ok(())
     }
 
+    /// Get the head and tip of the range we need to sync
+    async fn get_head_and_tip<DB: Database>(
+        &self,
+        db: &StageDB<'_, DB>,
+        stage_progress: u64,
+    ) -> Result<(H256, H256), StageError> {
+        let mut cursor = db.cursor::<tables::CanonicalHeaders>()?;
+
+        // Lookup the head hash and reposition the cursor to it
+        let (_, head) = cursor
+            .seek_exact(stage_progress)?
+            .ok_or(DatabaseIntegrityError::CanonicalHash { number: stage_progress })?;
+
+        let tip = match cursor.next()? {
+            Some((number, hash)) if stage_progress + 1 != number => hash,
+            None => self.next_fork_choice_state(&head).await.head_block_hash,
+            // if next header is present and there is no gap - checkpoint is incorrect
+            _ => return Err(StageError::Checkpoint(stage_progress)),
+        };
+
```

Comment:
trace here would be nice ("Syncing from x to y", "Filling gap between a and b")

Reply:
added in a different place